sampo/parser/general_build.py

```diff
@@ -0,0 +1,181 @@
+    """
+    The function find all elementary circuits using the algorithm of Donald B. Johnson
+    doi: 10.1137/0205007
+
+    :param works_info: dataframe, that contains information about works
+    :return: list of cycles
+    """
+    graph = nx.DiGraph()
+    edges = []
+
+    for _, row in works_info[::-1].iterrows():
+        v = row['activity_id']
+        for u in row['predecessor_ids']:
+            edges.append((v, u))
+
+    graph.add_nodes_from(list(works_info['activity_id']))
+    graph.add_edges_from(edges)
+
+    return list(nx.simple_cycles(graph))
+
+
+def fix_df_column_with_arrays(column: pd.Series, cast: Callable[[str], Any] | None = str,
+                              none_elem: Any | None = NONE_ELEM):
+    new_column = column.copy().astype(str).apply(
+        lambda elems: [cast(elem) for elem in elems.split(',')] if elems != str(math.nan) else [none_elem])
+    return new_column
+
+
+def preprocess_graph_df(frame: pd.DataFrame) -> pd.DataFrame:
+    def normalize_if_number(s):
+        return str(int(float(s))) \
+            if s.replace('.', '', 1).isdigit() \
+            else s
+
+    frame['activity_id'] = frame['activity_id'].astype(str)
+    frame['volume'] = frame['volume'].astype(float)
+    frame['predecessor_ids'] = fix_df_column_with_arrays(frame['predecessor_ids'], cast=normalize_if_number)
+    frame['connection_types'] = fix_df_column_with_arrays(frame['connection_types'],
+                                                          cast=EdgeType,
+                                                          none_elem=EdgeType.FinishStart)
+    if 'lags' not in frame.columns:
+        frame['lags'] = [NONE_ELEM] * len(frame)
+    frame['lags'] = fix_df_column_with_arrays(frame['lags'], float)
+
+    frame = break_circuits_in_input_work_info(frame)
+
+    return frame
+
+
+def add_graph_info(frame: pd.DataFrame) -> pd.DataFrame:
+    existed_ids = set(frame["activity_id"])
+
+    predecessor_ids, connection_types, lags = [], [], []
+    for _, row in frame[['predecessor_ids', 'connection_types', 'lags']].iterrows():
+        predecessor_ids.append([])
+        connection_types.append([])
+        lags.append([])
+        for index in range(len(row['predecessor_ids'])):
+            if row['predecessor_ids'][index] in existed_ids:
+                predecessor_ids[-1].append(row['predecessor_ids'][index])
+                connection_types[-1].append(row['connection_types'][index])
+                lags[-1].append(row['lags'][index])
+        if len(predecessor_ids[-1]) == 0:
+            predecessor_ids[-1].append(NONE_ELEM)
+            connection_types[-1].append(EdgeType.FinishStart)
+            lags[-1].append(float(NONE_ELEM))
+    frame['predecessor_ids'], frame['connection_types'], frame['lags'] = predecessor_ids, connection_types, lags
+
+    frame["edges"] = frame[['predecessor_ids', 'connection_types', 'lags']].apply(lambda row: list(zip(*row)), axis=1)
```

Comment:
replace all `"` with `'`
This style breakings should't appear at the same line...